File: 3518-smallest-palindromic-rearrangement-ii/3518-smallest-palindromic-rearrangement-ii.ts
function smallestPalindrome(s: string, k: number): string {
    const n = s.length;
    const half = n >> 1;
    const freq = new Array(26).fill(0);
    for (const ch of s) freq[ch.charCodeAt(0) - 97]++;

    const res: string[] = new Array(n).fill(' ');
    // middle character
    for (let i = 0; i < 26; i++) {
        if (freq[i] & 1) res[n >> 1] = String.fromCharCode(97 + i);
        freq[i] >>= 1;
    }

    const nCr = (n: number, r: number, limit: number): number => {
        r = Math.min(r, n - r);
        let ans = 1;
        for (let i = 1; i <= r; i++) {
            ans = ans * (n - i + 1) / i;
            if (ans > limit) return limit + 1;
        }
        return ans;
    };

    const ways = (rem: number, f: number[], limit: number): number => {
        let total = 1;
        for (let i = 0; i < 26; i++) {
            total *= nCr(rem, f[i], limit);
            if (total > limit) return limit + 1;
            rem -= f[i];
        }
        return total;
    };

    const total = ways(half, freq, k);
    if (total < k) return "";

    for (let idx = 0; idx < half; idx++) {
        for (let i = 0; i < 26; i++) {
            if (freq[i] === 0) continue;
            freq[i]--;
            const possible = ways(half - idx - 1, freq, k);
            if (possible >= k) {
                res[idx] = String.fromCharCode(97 + i);
                break;
            } else {
                k -= possible;
                freq[i]++;
            }
        }
    }

    for (let i = 0; i < half; i++) {
        res[n - 1 - i] = res[i];
    }
    return res.join('');
}

File: 3518-smallest-palindromic-rearrangement-ii/3518-smallest-palindromic-rearrangement-ii.rs
impl Solution {
    pub fn smallest_palindrome(s: String, mut k: i32) -> String {
        let n = s.len();
        let half = n / 2;
        let mut freq = [0; 26];
        for b in s.bytes() {
            freq[(b - b'a') as usize] += 1;
        }

        let mut res = vec![' '; n];
        // middle character
        for i in 0..26 {
            if (freq[i] & 1) == 1 {
                res[n / 2] = (b'a' + i as u8) as char;
            }
            freq[i] /= 2;
        }

        let ncr = |n: i32, mut r: i32, limit: i32| -> i32 {
            r = r.min(n - r);
            let mut ans = 1i64;
            for i in 1..=r {
                ans = ans * (n - i + 1) as i64 / i as i64;
                if ans > limit as i64 { return limit + 1; }
            }
            ans as i32
        };

        let ways = |rem: i32, f: &[i32], limit: i32| -> i32 {
            let mut total = 1i64;
            let mut rem = rem;
            for i in 0..26 {
                total *= ncr(rem, f[i], limit) as i64;
                if total > limit as i64 { return limit + 1; }
                rem -= f[i];
            }
            total as i32
        };

        let total = ways(half as i32, &freq, k);
        if total < k { return String::new(); }

        for idx in 0..half {
            for i in 0..26 {
                if freq[i] == 0 { continue; }
                freq[i] -= 1;
                let possible = ways((half - idx - 1) as i32, &freq, k);
                if possible >= k {
                    res[idx] = (b'a' + i as u8) as char;
                    break;
                } else {
                    k -= possible;
                    freq[i] += 1;
                }
            }
        }

        for i in 0..half {
            res[n - 1 - i] = res[i];
        }
        res.iter().collect()
    }
}